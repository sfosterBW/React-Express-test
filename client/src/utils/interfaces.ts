export interface Fruit {
  id: string
  name: string
  best: boolean
  description?: string
}

export type NewFruit = Omit<Fruit, 'id'>
