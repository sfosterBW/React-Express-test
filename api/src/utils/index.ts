import { IFruit, NewFruit } from './types'

/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment*/

const isBoolean = (bool: unknown): bool is boolean => {
  return typeof bool === 'boolean'
}

const isString = (text: unknown): text is string => {
  return typeof text === 'string' || text instanceof String;
}

const parseBest = (best: unknown): boolean => {
  if (typeof best === 'undefined' || !isBoolean(best)) {
    throw new Error(`Incorrect or missing best: ${String(best)}`);
  }

  return best
}

const parseName = (name: unknown): string => {
  if (!name || !isString(name)) {
    throw new Error(`Incorrect or missing name: ${String(name)}`)
  }

  return name
}

const parseDescription = (description: unknown): string => {
  if (!description || !isString(description)) {
    throw new Error(`Incorrect or missing description: ${String(description)}`)
  }

  return description
}

export const toNewFruit = (object: any): NewFruit => {
  const newFruit: NewFruit = {
    name: parseName(object.name),
    best: parseBest(object.best),
    description: object.description ? parseDescription(object.description) : undefined
  }

  return newFruit
}

const parseId = (id: unknown): string => {
  if (!id || !isString(id)) {
    throw new Error(`Incorrect or missing id: ${String(id)}`)
  }

  return id
}

export const toFruit = (object: any): IFruit => {
  const fruit: IFruit = {
    ...object,
    id: parseId(object.id),
    name: parseName(object.name),
    best: parseBest(object.best),
    description: object.description ? parseDescription(object.description) : undefined
  }

  return fruit
}
