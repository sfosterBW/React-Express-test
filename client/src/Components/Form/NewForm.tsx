import React, { FC, FormEvent } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { InputCheckbox, InputText } from '../Input/Input'
import useField from '../../utils/hooks'
import { RootState } from '../../Reducers/store'
import { toggleAlert } from '../../Reducers/alertReducer'
import { createFruit } from '../../Reducers/fruitReducer'
import { toggleModal } from '../../Reducers/modalReducer'
import styles from './Form.module.css'

const NewForm: FC<{ title?: string }> = ({ title = 'Add a new fruit' }) => {
  const dispatch = useDispatch()

  const modal = useSelector((state: RootState) => state.modal.toggle)

  const [best, resetBest] = useField<boolean>('Is it best?', 'best', 'checkbox', false)
  const [description, resetDescription] = useField<string>('Description', 'description', 'text', '')
  const [name, resetName] = useField<string>('Add a fruit', 'name', 'text', '')

  const submit = (event: FormEvent<HTMLFormElement>): void => {
    event.preventDefault()
    dispatch(createFruit({
      name: name.value,
      description: description.value === '' ? undefined : description.value,
      best: best.value
    }))
    dispatch(toggleAlert(`${name.value} has been added`, true))

    if (modal) {
      dispatch(toggleModal(false))
    }

    resetName()
    resetDescription()
    resetBest()
  }

  return (
    <form className={styles.form} onSubmit={(event): void => submit(event)}>
      <h2 className={styles.title} data-testid="new-form-title">
        {title}
      </h2>
      <InputText {...name} />
      <InputText {...description} />
      <InputCheckbox {...best} />
      <button className={styles.button} data-testid="new-form-submit">
        Add new fruit
      </button>
    </form>
  )
}

export default NewForm
