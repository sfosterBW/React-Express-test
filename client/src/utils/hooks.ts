import { ChangeEvent, useState } from 'react'

/* eslint-disable  @typescript-eslint/no-explicit-any */

interface Field<T> {
  label: string;
  name: string;
  onChange: (event: ChangeEvent<HTMLInputElement>) => void;
  type: string;
  value: T;
}

const useField = <T>(
  label: string,
  name: string,
  type: string,
  initValue: T
): [Field<T>, () => void] => {

  const [value, setValue] = useState(initValue)

  const onChange = (event: ChangeEvent<HTMLInputElement>): void => {
    if (type === 'checkbox' && typeof value === 'boolean') {
      setValue(event.target.checked as any)
    } else {
      setValue(event.target.value as any)
    }
  }

  const reset = (resetValue: T = initValue): void => {
    setValue(resetValue)
  }

  return [{
    onChange,
    label,
    name,
    type,
    value
  }, reset]
}

export default useField
