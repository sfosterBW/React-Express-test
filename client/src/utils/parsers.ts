import { Fruit } from './interfaces'

/* eslint-disable  @typescript-eslint/no-explicit-any */
/* eslint-disable  @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */

const isBoolean = (bool: unknown): bool is boolean => {
  return typeof bool === 'boolean'
}

const isString = (text: unknown): text is string => {
  return typeof text === 'string' || text instanceof String
}

const parseBest = (best: unknown): boolean => {
  if (typeof best === 'undefined' || !isBoolean(best)) {
    throw new Error(`Incorrect or missing best: ${String(best)}`)
  }

  return best
}

export const parseId = (id: unknown): string => {
  if (!id || !isString(id)) {
    throw new Error(`Incorrect or missing id: ${String(id)}`)
  }

  return id
}

const parseName = (name: unknown): string => {
  if (!name || !isString(name)) {
    throw new Error(`Incorrect or missing name: ${String(name)}`)
  }

  return name
}

const parseDescription = (description: unknown): string => {
  if (!description || !isString(description)) {
    throw new Error(`Incorrect or missing description: ${String(description)}`)
  }

  return description
}

const parseArray = (array: any): Array<unknown> => {
  if (!array || !Array.isArray(array)) {
    throw new Error(`Incorrect or missing an array ${String(array)}`)
  }

  return array
}

export const toFruit = (object: any): Fruit => {
  const fruit: Fruit = {
    id: parseId(object.id),
    name: parseName(object.name),
    best: parseBest(object.best)
  }

  if (object.description) {
    fruit.description = parseDescription(object.description)
  }

  return fruit
}

export const toFruits = (object: any): Fruit[] => {
  const array = parseArray(object)
  return array.map(fruit => toFruit(fruit))
}

export const toId = (object: any): string => {
  return parseId(object)
}

export default { toFruit, toFruits, toId }
