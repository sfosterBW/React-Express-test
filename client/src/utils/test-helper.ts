import { Fruit } from './interfaces'

export const emptyList: Fruit[] = []

export const fruit: Fruit = {
  id: "1",
  name:"apple",
  best:false,
  description: "I don't like doctors"
}

export const list: Fruit[] = [
  {
    id: "1",
    name: "apple",
    best: false,
    description: "I don't like doctors"
  },
  {
    id: "2",
    name: "banana",
    best: true
  }
]

export default { emptyList, fruit, list }
