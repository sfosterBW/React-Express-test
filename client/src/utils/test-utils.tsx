import React, { ReactNode } from 'react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import configureStore from "redux-mock-store"
import thunk from "redux-thunk"

import { initState as initAlertState } from '../Reducers/alertReducer'
import { initState as initFruitState } from '../Reducers/fruitReducer'
import { initState as initModalState } from '../Reducers/modalReducer'
import { RootState } from '../Reducers/store'

export const rootInitialState = {
  alert: initAlertState,
  fruit: initFruitState,
  modal: initModalState
};

export const renderWithProviders = (
  ui: ReactNode,
  initialState: RootState = rootInitialState
) => {
  const store = configureStore([thunk])(initialState)
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
    mockStore: store,
    getActionsTypes: () => store.getActions().map(a => a.type)
  }
}
