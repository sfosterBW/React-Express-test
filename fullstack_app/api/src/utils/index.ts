import { IFruit, NewFruit } from './types'

/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */

const isBoolean = (bool: any): bool is boolean => {
  return typeof bool === 'boolean'
}

const isString = (text: any): text is string => {
  return typeof text === 'string' || text instanceof String;
}

const parseBest = (best: any): boolean => {
  if (typeof best === 'undefined' || !isBoolean(best)) {
    throw new Error(`Incorrect or missing best: ${String(best)}`);
  }

  return best
}

const parseName = (name: any): string => {
  if (!name || !isString(name)) {
    throw new Error(`Incorrect or missing name: ${String(name)}`)
  }

  return name
}

export const toNewFruit = (object: any): NewFruit => {
  const newFruit: NewFruit = {
    ...object,
    name: parseName(object.name),
    best: parseBest(object.best)
  }

  return newFruit
}

const parseId = (id: any): string => {
  if (!id || !isString(id)) {
    throw new Error(`Incorrect or missing id: ${String(id)}`)
  }

  return id
}

export const toFruit = (object: any): IFruit => {
  const fruit: IFruit = {
    ...object,
    id: parseId(object.id),
    name: parseName(object.name),
    best: parseBest(object.best)
  }

  return fruit
}
