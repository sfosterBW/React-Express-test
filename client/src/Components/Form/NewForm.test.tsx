import React from 'react'
import { fireEvent } from '@testing-library/react'
import { renderWithProviders } from '../../utils/test-utils'

import NewForm from './NewForm'

describe('the Form component', () => {
  const component = <NewForm />

  it('functions properly with props', () => {
    const { getByLabelText, getByTestId } = renderWithProviders(component)
    const testName = 'Apple'
    const testDescription = 'I don\'t like doctors'

    fireEvent.click(getByTestId('toggle'))
    fireEvent.change(getByLabelText('Add a fruit'), {
      target: { value: testName }
    })
    expect(getByLabelText('Add a fruit')).toHaveValue(testName)

    fireEvent.change(getByLabelText('Description'), {
      target: { value: testDescription }
    })
    expect(getByLabelText('Description')).toHaveValue(testDescription)

    fireEvent.click(getByTestId('new-form-submit'))
    expect(getByLabelText('Add a fruit')).toHaveValue('')
    expect(getByLabelText('Description')).toHaveValue('')
  })

  it('renders the same as last time', () => {
    const { container } = renderWithProviders(component)
    expect(container).toMatchSnapshot()
  })
})
