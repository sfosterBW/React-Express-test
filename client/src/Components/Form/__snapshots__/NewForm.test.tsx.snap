// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Form component renders the same as last time 1`] = `
<div>
  <form
    class="form"
  >
    <h2
      class="title"
      data-testid="new-form-title"
    >
      Add a new fruit
    </h2>
    <div
      class="wrapper"
      data-testid="wrapper"
    >
      <label
        class="label"
        for="name"
      >
        Add a fruit
      </label>
      <input
        class="text"
        data-testid="input-text"
        id="name"
        name="name"
        type="text"
        value=""
      />
    </div>
    <div
      class="wrapper"
      data-testid="wrapper"
    >
      <label
        class="label"
        for="description"
      >
        Description
      </label>
      <input
        class="text"
        data-testid="input-text"
        id="description"
        name="description"
        type="text"
        value=""
      />
    </div>
    <div
      class="wrapper"
      data-testid="wrapper"
    >
      <label
        class="label"
        for="best"
      >
        Is it best?
      </label>
      <div
        class="checkboxWrapper"
      >
        <input
          class="checkbox"
          data-testid="input"
          id="best"
          name="best"
          type="checkbox"
          value="false"
        />
        <label
          class="checkboxLabel"
          data-testid="toggle"
          for="best"
        >
          <span
            class="checkboxButton"
          />
        </label>
      </div>
    </div>
    <button
      class="button"
      data-testid="new-form-submit"
    >
      Add new fruit
    </button>
  </form>
</div>
`;
